### ISSUE TITLE
Sticky Windows after suspend (or unplugging) for dual monitors

### ISSUE THREAD

**Issue by jean** [linuxmint/cinnamon#5371]
- I plug in an external display, and open various windows on both laptop and external displays.
- Unplug and re-plug external display. 
- Now all my windows on all workspaces are back on my laptop display. 

This means I need to manually drag them back over to the external display again.

Alternatively:
- I plug in an external display, and open various windows on both laptop and external displays.
- I suspend and wake up the laptop. 
- Now all my windows on all workspaces are on my external display. 

It would be great if the placement of windows on displays was remembered.

The original [forum discussion](https://forums.linuxmint.com/viewtopic.php?f=208&t=213331) mentions that the same behaviour occurs after the screen saver activates. I did experience this once or twice, but it went away again. 

System: Host: X250 Kernel: 3.19.0-32-generic x86_64 (64 bit)
Desktop: Cinnamon 2.8.8 Distro: Linux Mint 17.3 Rosa

**Comment by JosephMcc** [comment]
Your issue title doesn't match the description. The title says an issue when you use suspend. The description says the issue is with plugging/unplugging monitors. Those are two different things.

**Comment by jean** [comment]
@JosephMcc that's true. For me they tend to coincide, i.e. I always unplug the external display before suspending when I leave work. 

To clarify the issue, I tried it the other way round now. The result is that instead of all windows being shunted to the laptop display, they now all get shunted to the external display. The effect is the same: to restore my working environment, I need to move affected windows back to the other display manually.

**Comment by JosephMcc** [comment]
Thanks for clarification. The issue of windows moving to one monitor when the other is unplugged and not being restored when plugging it back in is something we won't fix. It really isn't broken in the first place. The windows should move to the connected monitors when one is unplugged. Trying to somehow keep track of where they were previously located on a removed monitor and keeping that info stored in memory just in case you plug it back in is not something we're going to do.

The  issue with all the windows moving to the primary when returning from suspend shouldn't happen though. In my own experience at least, issues with suspend seem to have a lot to do with the graphics hardware and driver on Cinnamon.

**Comment by jean** [comment]
> It really isn't broken in the first place.

It certainly is ... :stuck_out_tongue_winking_eye: 

A desktop keeps track of window layouts and configuration. That's its job. It keeps track of where everything is placed, which windows are sticky, which are always on top, which don't take focus --- and logically it should do this across displays. In fact, the more display area there is to manage, the more useful and necessary this task becomes. 

> The windows should move to the connected monitors when one is unplugged. 

Yes, that's logical too.

> somehow keep track of where they were previously located on a removed monitor and keeping that info stored in memory just in case you plug it back in

We don't care about the identify of the monitor, we just care about the field of pixels. 
I.e. I think another way of saying "display is (un)plugged" is "virtual screen size changes".

A possible discipline might be:
- when virtual screen size changes, 
  - store the pixel position on the virtual desktop of all windows that need to move;
  - try to move all windows with stored positions back to the stored position, taking into account the new size of the virtual screen. 

This way, windows would remember the last external display they were on. A refinement would be to allow saving a session. Then, the algorithm becomes, when a display is plugged in: 
- try to move to session position,
- fall back to previous external position,
- adjust to available space.

But I think remembering the last external display position would be great already. In my experience, a laptop is used with the same external display for long stretches of time.

**Comment by mainmachine** [comment]
At the risk of just being another +1, I too have been experiencing this issue:
I use a Dell Laptop with  Ryzen 5 SOC (so amdgpu/ mesa graphics stack) with an external HDMI connected display. In typical use I will have windows open on both the external and internal displays. I walk away from my desk, screensaver kicks in, then when I unlock all of my windows are on the internal display.
Is there anything I can do to assist in debugging this?
I value cinnamon and the work that goes into it, and would love to help contribute back.

**Comment by mstahl302** [comment]
For the duel monitor / sleep issue, I've been looking for workarounds. I've found two on the web mentioned frequently.  You may want to try them. 

Disable xrandr (a window plugin)

`dconf read /org/cinnamon/settings-daemon/plugins/xrandr/active false`

Disable monitor sleep mode

Settings > Power Management > "Turn off the screen when inactive" > Never

Good luck!
